### PR TITLE
core: append LogExtraFields= values to log_unit* messages

### PR DESCRIPTION
**Pull request by bluca**

### CODE REVIEW

**Comment by DaanDeMeyer** [comment]
This works, but the whole point of the context was that it was independent of the logging macros. So instead of putting it in the log_unit() macros. I would append LogExtraFields to the log context at the start of every callback that is related to a unit.

**Comment by bluca** [comment]
> This works, but the whole point of the context was that it was independent of the logging macros. So instead of putting it in the log_unit() macros. I would append LogExtraFields to the log context at the start of every callback that is related to a unit.

Which callbacks are those? Could you please link an example?

**Comment by DaanDeMeyer** [comment]
> Which callbacks are those? Could you please link an example?

Callbacks is the wrong word. The idea is that whenever you start executing code that operates on a specific unit, you'd push that units `LogExtraFields` iovec onto the log context. You'd add some helper `LOG_CONTEXT_PUSH_UNIT()` or so and then call that as `LOG_CONTEXT_PUSH_UNIT(u)` whenever we start operating on a unit. So everywhere in the code when we start doing something with a unit object, you'd call that and the fields will get added to every log message logged in that block.

**Comment by bluca** [comment]
But that requires massive changes, no? What's the issue with this approach? End result is the same?

**Comment by bluca** [comment]
There's almost 1k log_unit calls right now:

```
$ git grep -r log_unit src | wc -l
980
```

I really don't want to go and add 1k changes. Also, every time a new log_unit call is added anywhere, we'd need to manually ensure the same change is also applied while reviewing. And if a block does multiple log_unit calls about different units, it would need to be split into different blocks. I'm not sure this would be the best approach? What are the upsides?

**Comment by DaanDeMeyer** [comment]
> I really don't want to go and add 1k changes. Also, every time a new log_unit call is added anywhere, we'd need to manually ensure the same change is also applied while reviewing. And if a block does multiple log_unit calls about different units, it would need to be split into different blocks. I'm not sure this would be the best approach? What are the upsides?

The idea is that you call LOG_CONTEXT_PUSH_UNIT() in a few tactically chosen locations and it will automatically apply to any new logging calls that are added in the future. e.g. if you add a call in manager_invoke_sigchld_event() in src/core/manager.c, every logging call made in that callback will have the fields added. `bus_unit_method_start_generic()` would also be a good place to put one.

It's not 100% trivial to figure out where to retroactively add all the log context pushes though, I'll give you that.

I'm fine with merging this since the goal is to augment the existing unit log messages and not apply the fields to new ones, but it's not quite using the log context in the intended way.

**Comment by bluca** [comment]
The thing is that even if we find all the right locations as things stand now, we will need to manually keep it updated as we go along, and there's no way to ensure that all log messages of units include these fields using that solution. That's what I want to achieve: ensure that whatever happens, these fields are included, and we don't have to keep updating it by hand, as that's just cumbersome. If there's a better way to achieve this that's fine by me, but it really needs to work out implicitly and automatically.

**Comment by DaanDeMeyer** [comment]
> The thing is that even if we find all the right locations as things stand now, we will need to manually keep it updated as we go along, and there's no way to ensure that all log messages of units include these fields using that solution. That's what I want to achieve: ensure that whatever happens, these fields are included, and we don't have to keep updating it by hand, as that's just cumbersome. If there's a better way to achieve this that's fine by me, but it really needs to work out implicitly and automatically.

It's not ideal either way, either you have to make sure you use log_unit_xxx() everywhere or you have to make sure you push onto the context everywhere required. But since we already use log_unit_xxx() everywhere, you'll probably have better coverage this way.

**Review by DaanDeMeyer — APPROVED** [review]

File: src/basic/log.c [hunk 1]
```diff
@@ -613,14 +615,23 @@ static void log_do_context(struct iovec *iovec, size_t iovec_len, size_t *n) {
...
 
                         iovec[(*n)++] = IOVEC_MAKE_STRING(*s);
                         iovec[(*n)++] = IOVEC_MAKE_STRING("\n");
                 }
+
+                for (size_t i = 0; i < c->n_input_iovec; i++) {
+                        if (*n + 2 >= iovec_len)
+                                return;
+
+                        iovec[(*n)++] = IOVEC_MAKE(c->input_iovec[i].iov_base, c->input_iovec[i].iov_len);
```
Comment (poettering) [inline src/basic/log.c]: wouldn't this be more obvious:

```c
iovec[(*n++)] = c->input_iovec[i];
```

File: src/core/unit.h [hunk 2]
```diff
@@ -1078,6 +1078,15 @@ Condition *unit_find_failed_condition(Unit *u);
...
+                struct iovec *_input_iovec = NULL;                      \
+                size_t _n_input_iovec = 0;                              \
+                const ExecContext *_c = _u ?                            \
+                        unit_get_exec_context(_u) : NULL;               \
+                if (_c && unit_log_level_test(_u, _l)) {                \
+                        _input_iovec = _c->log_extra_fields;            \
+                        _n_input_iovec = _c->n_log_extra_fields;        \
+                }                                                       \
+                LOG_CONTEXT_PUSH_IOV(_input_iovec, _n_input_iovec);     \
                 (log_get_max_level() < LOG_PRI(_l) || (_u && !unit_log_level_test(_u, _l))) ? -ERRNO_VALUE(error) : \
```
Comment (poettering) [inline src/core/unit.h]: this doesn't look right. You do the unit_log_level_test() twice, but not the log_get_max_level(). 

Please do both, only once, and do not invoke unit_get_exec_context() unless they pass.

File: src/basic/log.c (hunk 1 above)
Comment (bluca) [reply src/basic/log.c]: less aesthetically pleasing though :-P

**Comment by bluca** [comment]
Assuming this is transient, could you please restart the job for me @mrc0mmand 

```
16:48:18 Errors during downloading metadata for repository 'centos-kmods-source':
16:48:18   - Status code: 403 for https://vault.centos.org/centos/8-stream/kmods/Source/packages-main/repodata/repomd.xml (IP: 18.165.83.95)
16:48:18 Error: Failed to download metadata for repo 'centos-kmods-source': Cannot download repomd.xml: Cannot download repodata/repomd.xml: Al
```

**Pull request merged**